xapian-applications/omega/handler_libextractor.cc

```diff
@@ -0,0 +1,136 @@
+/** @file handler_libextractor.cc
+ * @brief Extract metadata using libextractor.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <extractor.h>
+#include <cstring>
+
+using namespace std;
+
+struct metadata {
+    string& title;
+    string& author;
+    string& keywords;
+    string& pages;
+
+    metadata(string& t, string& a, string& k, string& p)
+	    : title(t), author(a), keywords(k), pages(p) {}
+};
+
+/*
+* cls: passed as last parameter from EXTRACTOR_extract
+* plugin_name: name of the plugin
+* type: mime-type of file according to libextractor
+* format: format information about data
+* data_mime_type: mimetype of data according to libextractor
+* data: actual meta-data found
+* data_len: number of bytes in data
+*/
+static int
+process_metadata(void* cls,
+		 const char* plugin_name,
+		 enum EXTRACTOR_MetaType type,
+		 enum EXTRACTOR_MetaFormat format,
+		 const char* data_mime_type,
+		 const char* data,
+		 size_t data_len)
+{
+    struct metadata* md = (struct metadata*)cls;
```

Comment:
Better to avoid C-style casts in C++ code.  A C-style cast can be used for one of several purposes and it may not be clear to the reader which are wanted.  Also the compiler doesn't know which, and that allows errors to slip through.  C++ casts make the purpose clear to the reader, and the compiler can complain if they're misused.  `static_cast<>` is the appropriate one here: https://en.cppreference.com/w/cpp/language/static_cast

As an example of the sort of problem this avoids, if we had `const void* cls2` and tried:

```
    struct metadata* md = (struct metadata*)cls2;
```

Then this would compile without even a warning (because one use of a C-style cast is to remove `const` qualification), and we'd be able to modify the object despite getting a `const` pointer to it - that's potentially very bad.  If the object is actually in the `RODATA` section, it could give an access violation.  If it's constant data but not protected in that way, it's undefined behaviour to modify it.  Even if it's not constant, someone reading code that passes in an object as a `const` pointer won't expect it to be modified, and reasoning about calling code may be invalid.

If you try to compile this you'll get an error:

```
    struct metadata* md = static_cast<struct metadata*>(cls2);
```

If you actually want to remove `const` with a C++ cast you need to use `const_cast<>` (and that only removes `const` - it won't accidentally change the type at the same time).

Reply:
Thank you for your insight. I have changed this.